smart-contracts/advanced/multicall.md

```diff
@@ -0,0 +1,141 @@
+
+1. Solidity Implementation:
+   The `aggregate3` method is implemented in the Multicall3 contract like this:
+
+   ```solidity
+   function aggregate3(Call3[] calldata calls) public payable returns (Result[] memory returnData) {
+       uint256 length = calls.length;
+       returnData = new Result[](length);
+       for (uint256 i = 0; i < length;) {
+           (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+           if (calls[i].allowFailure) {
+               returnData[i] = Result(success, ret);
+           } else {
+               require(success, "Multicall3: call failed");
+               returnData[i] = Result(true, ret);
+           }
+           unchecked { ++i; }
+       }
+   }
+   ```
+
+2. Example of sending multicalls to this smart contract:
+   Here's an example using ethers.js to interact with the Multicall3 contract:
+
+   ```javascript
+   const { ethers } = require("ethers");
+
+   const provider = new ethers.providers.JsonRpcProvider("https://api.node.glif.io/rpc/v1");
+   const multicallAddress = "0xcA11bde05977b3631167028862bE2a173976CA11";
+   const multicallAbi = [/* Multicall3 ABI */];
+   const multicall = new ethers.Contract(multicallAddress, multicallAbi, provider);
+
+   // Example: Batch balance checks for multiple addresses
+   async function batchBalanceChecks(addresses) {
+     const calls = addresses.map(address => ({
+       target: "0x...", // ERC20 token address
+       allowFailure: false,
+       callData: ethers.utils.id("balanceOf(address)").slice(0, 10) + 
+                 ethers.utils.defaultAbiCoder.encode(["address"], [address]).slice(2)
+     }));
+
+     const results = await multicall.aggregate3(calls);
+     return results.map(result => ethers.utils.defaultAbiCoder.decode(["uint256"], result.returnData)[0]);
+   }
+
+   batchBalanceChecks(["0x123...", "0x456...", "0x789..."]).then(console.log);
+   ```
+
+This example demonstrates how to use Multicall3 to batch multiple `balanceOf` calls for an ERC20 token in a single transaction, significantly reducing the number of separate RPC calls needed.
+
+### Batch Contract Writes
+
+_If using Multicall3 for this purpose, be aware it is unaudited, so use at your own risk._
+_However, because it is a stateless contract, it should be safe when used correctly—**it should never hold your funds after a transaction ends, and you should never approve Multicall3 to spend your tokens**_.
+
+Multicall3 can also be used to batch on-chain transactions using the methods described in the [Batch Contract Reads](#batch-contract-reads) section.
+
+When using Multicall3 for this purpose, there are **two important details you MUST understand**.
+
+1. How `msg.sender` behaves when calling vs. delegatecalling to a contract.
+2. The risks of using `msg.value` in a multicall.
+
+Before explaining both of these, let's first cover some background.
+
+There are two types of accounts: Externally Owned Accounts (EOAs) and Contract Accounts.
```

Comment:
This whole part of explaining EOA & Contract, call & delegateCall is very useful. But it feels too heavy for our docs here. Is it possible to simplify it to a couple of paragraphs?

Reply:
I tried shortening it a bit, but I do think that there is very important information toward the bottom about how Multicall3 is used in practice and what to / not to rely on from a functionality and security perspective; let me know what you think!

Reply:
Resolved, shortened